### ISSUE TITLE
product.update webhook fires 3 times; fails with V3

### ISSUE THREAD

**Issue by LaurenaRehbein** [woocommerce/woocommerce#21867]
**Describe the bug**
When a `product.update` webhook is created, and a product is updated, the webhook fires three times. 

**To Reproduce**
1. Create a webhook for Product Updated and save, using API V2 
2. Update a product
3. See that the hook fired 3 times
4. View Logs to confirm fired multiple times
5. View Scheduled Actions to see triplicate entries

**Screenshots**

![http://cld.wthms.co/WKUuXL](http://cld.wthms.co/WKUuXL+)
Full Image: http://cld.wthms.co/WKUuXL+ 

**Expected behavior**
I expected this to fire one time. 

Next, I added a same webhook but used API V3 and repeated the test. Also fired 3 times, but failed, with this error: 

Error: 

```
2018-11-09T00:33:05+00:00 CRITICAL Uncaught Error: Class 'WC_REST_Products_V3_Controller' not found in /wc-test/wp-content/plugins/woocommerce/includes/class-wc-webhook.php:297
Stack trace:
#0 /wc-test/wp-content/plugins/woocommerce/includes/class-wc-webhook.php(349): WC_Webhook->get_wp_api_payload('product', 564, 'updated')
#1 /wc-test/wp-content/plugins/woocommerce/includes/class-wc-webhook.php(187): WC_Webhook->build_payload(564)
#2 /wc-test/wp-content/plugins/woocommerce/includes/wc-webhook-functions.php(46): WC_Webhook->deliver(564)
#3 /wordpress/core/4.9.8/wp-includes/class-wp-hook.php(286): wc_deliver_webhook_async(5, 564)
#4 /wordpress/core/4.9.8/wp-includes/class-wp-hook.php(310): WP_Hook->apply_filters('', Array)
#5 /wordpress/core/4.9.8/wp-includes/plugin.php(515): WP_Hook->do_action(Array)
#6 /wc-test/wp-content/plugins/woocommerce/includes/libraries/action-scheduler/classes/ActionScheduler_Action.php(22): do_action_ref_array('woocommerce_del...', Array)
#7 /wc-test/wp-content/plugins/woocommerce/includes/libraries/action-scheduler/classes/ActionScheduler_Abstract_QueueRunner.php(59): ActionScheduler_Action->execute()
#8 /wc-test/wp-content/plugins/woocommerce/includes/libraries/action-scheduler/classes/ActionScheduler_QueueRunner.php(80): ActionScheduler_Abstract_QueueRunner->process_action('578')
#9 /wc-test/wp-content/plugins/woocommerce/includes/libraries/action-scheduler/classes/ActionScheduler_QueueRunner.php(61): ActionScheduler_QueueRunner->do_batch(25)
#10 /wordpress/core/4.9.8/wp-includes/class-wp-hook.php(286): ActionScheduler_QueueRunner->run()
#11 /wordpress/core/4.9.8/wp-includes/class-wp-hook.php(310): WP_Hook->apply_filters('', Array)
#12 /wordpress/core/4.9.8/wp-includes/plugin.php(515): WP_Hook->do_action(Array)
#13 /wordpress/core/4.9.8/wp-cron.php(126): do_action_ref_array('action_schedule...', Array)
#14 {main}
  thrown in /wc-test/wp-content/plugins/woocommerce/includes/class-wc-webhook.php on line 297
```

Reported in 1550087-zen and replicated on a test site. 

**Isolating the problem (mark completed items with an [x]):**
- [X] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [X] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/storefront/).
- [X] I can reproduce this bug consistently using the steps above.

**WordPress Environment**
<details>
```
`
### WordPress Environment ###

Home URL: https://wc-test.mystagingwebsite.com
Site URL: https://wc-test.mystagingwebsite.com
WC Version: 3.5.1
Log Directory Writable: ✔
WP Version: 4.9.8
WP Multisite: –
WP Memory Limit: 256 MB
WP Debug Mode: ✔
WP Cron: ✔
Language: en_US
External object cache: ✔

### Server Environment ###

Server Info: nginx
PHP Version: 7.0.32 - We recommend using PHP version 7.2 or above for greater performance and security. How to update your PHP version
PHP Post Max Size: 100 MB
PHP Time Limit: 300
PHP Max Input Vars: 6144
cURL Version: 7.61.1
OpenSSL/1.0.1t

SUHOSIN Installed: –
MySQL Version: 5.5.5-10.1.17-MariaDB
Max Upload Size: 100 MB
Default Timezone is UTC: ✔
fsockopen/cURL: ✔
SoapClient: ✔
DOMDocument: ✔
GZip: ✔
Multibyte String: ✔
Remote Post: ✔
Remote Get: ✔

### Database ###

WC Database Version: 3.4.5
WC Database Prefix: wp_
MaxMind GeoIP Database: ✔
Total Database Size: 9.62MB
Database Data Size: 5.59MB
Database Index Size: 4.03MB
wp_woocommerce_sessions: Data: 0.02MB + Index: 0.02MB
wp_woocommerce_api_keys: Data: 0.02MB + Index: 0.03MB
wp_woocommerce_attribute_taxonomies: Data: 0.02MB + Index: 0.02MB
wp_woocommerce_downloadable_product_permissions: Data: 0.02MB + Index: 0.05MB
wp_woocommerce_order_items: Data: 0.02MB + Index: 0.02MB
wp_woocommerce_order_itemmeta: Data: 0.08MB + Index: 0.09MB
wp_woocommerce_tax_rates: Data: 0.02MB + Index: 0.06MB
wp_woocommerce_tax_rate_locations: Data: 0.02MB + Index: 0.03MB
wp_woocommerce_shipping_zones: Data: 0.02MB + Index: 0.00MB
wp_woocommerce_shipping_zone_locations: Data: 0.02MB + Index: 0.03MB
wp_woocommerce_shipping_zone_methods: Data: 0.02MB + Index: 0.00MB
wp_woocommerce_payment_tokens: Data: 0.02MB + Index: 0.02MB
wp_woocommerce_payment_tokenmeta: Data: 0.02MB + Index: 0.03MB
wp_woocommerce_log: Data: 0.02MB + Index: 0.02MB
wpstg1_commentmeta: Data: 0.02MB + Index: 0.03MB
wpstg1_comments: Data: 0.02MB + Index: 0.09MB
wpstg1_followup_coupons: Data: 0.02MB + Index: 0.03MB
wpstg1_followup_coupon_logs: Data: 0.02MB + Index: 0.03MB
wpstg1_followup_customers: Data: 0.02MB + Index: 0.06MB
wpstg1_followup_customer_carts: Data: 0.02MB + Index: 0.03MB
wpstg1_followup_customer_notes: Data: 0.02MB + Index: 0.03MB
wpstg1_followup_customer_orders: Data: 0.02MB + Index: 0.02MB
wpstg1_followup_email_coupons: Data: 0.02MB + Index: 0.02MB
wpstg1_followup_email_excludes: Data: 0.02MB + Index: 0.05MB
wpstg1_followup_email_logs: Data: 0.02MB + Index: 0.05MB
wpstg1_followup_email_orders: Data: 0.02MB + Index: 0.09MB
wpstg1_followup_email_order_coupons: Data: 0.02MB + Index: 0.02MB
wpstg1_followup_email_tracking: Data: 0.02MB + Index: 0.08MB
wpstg1_followup_followup_history: Data: 0.02MB + Index: 0.05MB
wpstg1_followup_order_categories: Data: 0.02MB + Index: 0.03MB
wpstg1_followup_order_items: Data: 0.02MB + Index: 0.05MB
wpstg1_followup_subscribers: Data: 0.02MB + Index: 0.03MB
wpstg1_followup_subscribers_to_lists: Data: 0.02MB + Index: 0.03MB
wpstg1_followup_subscriber_lists: Data: 0.02MB + Index: 0.03MB
wpstg1_links: Data: 0.02MB + Index: 0.02MB
wpstg1_options: Data: 0.25MB + Index: 0.05MB
wpstg1_postmeta: Data: 0.19MB + Index: 0.19MB
wpstg1_posts: Data: 0.11MB + Index: 0.06MB
wpstg1_rg_form: Data: 0.02MB + Index: 0.00MB
wpstg1_rg_form_meta: Data: 0.02MB + Index: 0.00MB
wpstg1_rg_form_view: Data: 0.02MB + Index: 0.03MB
wpstg1_rg_incomplete_submissions: Data: 0.02MB + Index: 0.02MB
wpstg1_rg_lead: Data: 0.02MB + Index: 0.03MB
wpstg1_rg_lead_detail: Data: 0.02MB + Index: 0.06MB
wpstg1_rg_lead_detail_long: Data: 0.02MB + Index: 0.00MB
wpstg1_rg_lead_meta: Data: 0.02MB + Index: 0.05MB
wpstg1_rg_lead_notes: Data: 0.02MB + Index: 0.03MB
wpstg1_termmeta: Data: 0.02MB + Index: 0.03MB
wpstg1_terms: Data: 0.02MB + Index: 0.03MB
wpstg1_term_relationships: Data: 0.02MB + Index: 0.02MB
wpstg1_term_taxonomy: Data: 0.02MB + Index: 0.03MB
wpstg1_usermeta: Data: 0.02MB + Index: 0.03MB
wpstg1_users: Data: 0.02MB + Index: 0.05MB
wpstg1_woocommerce_api_keys: Data: 0.02MB + Index: 0.03MB
wpstg1_woocommerce_attribute_taxonomies: Data: 0.02MB + Index: 0.02MB
wpstg1_woocommerce_bundled_itemmeta: Data: 0.02MB + Index: 0.03MB
wpstg1_woocommerce_bundled_items: Data: 0.02MB + Index: 0.03MB
wpstg1_woocommerce_downloadable_product_permissions: Data: 0.02MB + Index: 0.05MB
wpstg1_woocommerce_log: Data: 0.02MB + Index: 0.02MB
wpstg1_woocommerce_order_itemmeta: Data: 0.02MB + Index: 0.03MB
wpstg1_woocommerce_order_items: Data: 0.02MB + Index: 0.02MB
wpstg1_woocommerce_payment_tokenmeta: Data: 0.02MB + Index: 0.03MB
wpstg1_woocommerce_payment_tokens: Data: 0.02MB + Index: 0.02MB
wpstg1_woocommerce_sessions: Data: 0.02MB + Index: 0.02MB
wpstg1_woocommerce_shipping_zones: Data: 0.02MB + Index: 0.00MB
wpstg1_woocommerce_shipping_zone_locations: Data: 0.02MB + Index: 0.03MB
wpstg1_woocommerce_shipping_zone_methods: Data: 0.02MB + Index: 0.00MB
wpstg1_woocommerce_tax_rates: Data: 0.02MB + Index: 0.06MB
wpstg1_woocommerce_tax_rate_locations: Data: 0.02MB + Index: 0.03MB
wp_ariadminer_connections: Data: 0.02MB + Index: 0.00MB
wp_commentmeta: Data: 0.02MB + Index: 0.03MB
wp_comments: Data: 0.13MB + Index: 0.09MB
wp_crm_log: Data: 0.02MB + Index: 0.00MB
wp_crm_log_meta: Data: 0.02MB + Index: 0.00MB
wp_links: Data: 0.02MB + Index: 0.02MB
wp_options: Data: 1.27MB + Index: 0.06MB
wp_postmeta: Data: 1.42MB + Index: 0.45MB
wp_posts: Data: 0.11MB + Index: 0.06MB
wp_rg_form: Data: 0.02MB + Index: 0.00MB
wp_rg_form_meta: Data: 0.02MB + Index: 0.00MB
wp_rg_form_view: Data: 0.02MB + Index: 0.03MB
wp_rg_incomplete_submissions: Data: 0.02MB + Index: 0.02MB
wp_rg_lead: Data: 0.02MB + Index: 0.03MB
wp_rg_lead_detail: Data: 0.02MB + Index: 0.06MB
wp_rg_lead_detail_long: Data: 0.02MB + Index: 0.00MB
wp_rg_lead_meta: Data: 0.02MB + Index: 0.05MB
wp_rg_lead_notes: Data: 0.02MB + Index: 0.03MB
wp_snippets: Data: 0.02MB + Index: 0.00MB
wp_termmeta: Data: 0.02MB + Index: 0.03MB
wp_terms: Data: 0.02MB + Index: 0.03MB
wp_term_relationships: Data: 0.02MB + Index: 0.02MB
wp_term_taxonomy: Data: 0.02MB + Index: 0.03MB
wp_usermeta: Data: 0.05MB + Index: 0.03MB
wp_users: Data: 0.02MB + Index: 0.05MB
wp_wc_booking_relationships: Data: 0.02MB + Index: 0.03MB
wp_wc_download_log: Data: 0.02MB + Index: 0.03MB
wp_wc_webhooks: Data: 0.02MB + Index: 0.02MB
wp_woocommerce_bundled_itemmeta: Data: 0.02MB + Index: 0.03MB
wp_woocommerce_bundled_items: Data: 0.02MB + Index: 0.03MB
wp_wps_hit: Data: 0.02MB + Index: 0.14MB
wp_wps_index: Data: 0.06MB + Index: 0.05MB
wp_wps_key: Data: 0.02MB + Index: 0.02MB
wp_wps_object_type: Data: 0.02MB + Index: 0.08MB
wp_wps_query: Data: 0.02MB + Index: 0.02MB
wp_wps_uri: Data: 0.02MB + Index: 0.02MB
wp_wps_user_agent: Data: 0.02MB + Index: 0.02MB

### Post Type Counts ###

attachment: 17
course: 1
custom_css: 2
follow_up_email: 2
global_product_addon: 1
jetpack_migration: 2
jp_img_sitemap: 1
jp_sitemap: 1
jp_sitemap_master: 1
nav_menu_item: 11
page: 21
post: 1
product: 14
product_variation: 50
revision: 26
scheduled-action: 3
shop_coupon: 1
shop_order: 55
shop_subscription: 1
wc_booking: 2
wc_membership_plan: 2
wishlist: 2

### Security ###

Secure connection (HTTPS): ✔
Hide errors from visitors: ✔

### Active Plugins (2) ###

WooCommerce Stripe Gateway: by WooCommerce – 4.1.12
WooCommerce: by Automattic – 3.5.1

### Settings ###

API Enabled: –
Force SSL: –
Currency: CAD ($)
Currency Position: left
Thousand Separator: ,
Decimal Separator: .
Number of Decimals: 2
Taxonomies: Product Types: accommodation-booking (accommodation-booking)
booking (booking)
bundle (bundle)
composite (composite)
external (external)
grouped (grouped)
photography (photography)
simple (simple)
subscription (subscription)
variable (variable)
variable subscription (variable-subscription)

Taxonomies: Product Visibility: exclude-from-catalog (exclude-from-catalog)
exclude-from-search (exclude-from-search)
featured (featured)
outofstock (outofstock)
rated-1 (rated-1)
rated-2 (rated-2)
rated-3 (rated-3)
rated-4 (rated-4)
rated-5 (rated-5)


### WC Pages ###

Shop base: #3 - /shop/
Cart: #4 - /cart/
Checkout: #5 - /checkout/
My account: #6 - /my-account/
Terms and conditions: ❌ Page not set

### Theme ###

Name: Storefront
Version: 2.3.5
Author URL: https://woocommerce.com/
Child Theme: ❌ – If you are modifying WooCommerce on a parent theme that you did not build personally we recommend using a child theme. See: How to create a child theme
WooCommerce Support: ✔

### Templates ###

Overrides: –
`
```
</details>

**Comment by mikkamp** [comment]
I'm able to replicate this as well with WooCommerce 3.5.1.

Regarding the issue of it not finding `WC_REST_Products_V3_Controller`. This is because for the latest version of the API we should be looking for `WC_REST_Products_Controller` instead. This comes from the following line of code: https://github.com/woocommerce/woocommerce/blob/3.5.1/includes/class-wc-webhook.php#L288

The if statement seems to be the wrong way around, so it only adds the suffix if we are on v3 and not when we are on lower versions.

Regarding the issue with the webhook firing 3 times. It seems to be hooked to the following 3 actions:

- woocommerce_process_product_meta
- woocommerce_update_product
- woocommerce_update_product_variation

For a simple product we don't update variations but we can still see it fired 3 times (an extra time for saving product images). Is it intended to fire multiple times if more then one of those actions is called? On my test site I can see it fired 3 times from the following locations:

```
1. do_action( 'woocommerce_process_product_meta' ) called at [woocommerce/includes/admin/class-wc-admin-meta-boxes.php:220]
2. WC_Product->save() called at [woocommerce/includes/admin/meta-boxes/class-wc-meta-box-product-data.php:394]
3. WC_Product->save() called at [woocommerce/includes/admin/meta-boxes/class-wc-meta-box-product-images.php:94]
```

Is there any way the webhook can be triggered to fire only once?

**Comment by kloon** [comment]
Class not found error was fixed in #21916

**Comment by Trischan** [comment]
Is it possible this issue is linked to the customer.created firing three times?
